oneflow/core/job_rewriter/auto_parallel.cpp

```diff
@@ -0,0 +1,153 @@
+               || op_conf.user_conf().op_type_name() == "hierarchical_parallel_cast"
+               || op_conf.user_conf().op_type_name() == "hierarchical_parallel_cast_like");
+  };
+  std::vector<std::string> del_op_names;
+  HashSet<std::string> del_op_name_set;
+  std::function<void(const OpNode*)> Try2Delete = [&](const OpNode* op_node) {
+    if (del_op_name_set.find(op_node->op().op_name()) != del_op_name_set.end()) { return; }
+    const OperatorConf& op_conf = op_node->op().op_conf();
+    if (!op_conf.ctrl_in_op_name().empty()) { return; }
+    if (ctrl_in_op_names.find(op_conf.name()) != ctrl_in_op_names.end()) { return; }
+    if (!IsParallelCastOp(op_conf)) { return; }
+    if (op_node->in_edges().size() != 1) { return; }
+
+    // Find the first op which won't be deleted
+    const OpNode* source_op = op_node;
+    const OpNode* producer = op_node->SoleInEdge()->src_node();
+    while (IsParallelCastOp(producer->op().op_conf())) {
+      Try2Delete(producer);
+      if (del_op_name_set.find(producer->op().op_name()) == del_op_name_set.end()) { break; }
+      source_op = producer;
+      producer = source_op->SoleInEdge()->src_node();
+    }
+    user_op::UserOpConfWrapper conf_wrapper_in(source_op->op().op_conf());
+    const LogicalBlobId& parallel_cast_in_lbi = GenLogicalBlobId(conf_wrapper_in.input("in", 0));
+
+    user_op::UserOpConfWrapper conf_wrapper_out(op_conf);
+    const LogicalBlobId& parallel_cast_out_lbi =
+        GenLogicalBlobId(conf_wrapper_out.output("out", 0));
+    if (op_node->parallel_desc() != producer->parallel_desc()) { return; }
+    for (const OpEdge* out_edge : op_node->out_edges()) {
+      const OpNode* consumer = out_edge->dst_node();
+      if (consumer->parallel_desc() != op_node->parallel_desc()) { return; }
+    }
+    op_name2nd_sbp_signature[producer->op().op_name()] = producer->nd_sbp_signature();
+    for (const OpEdge* out_edge : op_node->out_edges()) {
+      const OpNode* consumer = out_edge->dst_node();
+      const std::string& consumer_op_name = consumer->op().op_name();
+      op_name2nd_sbp_signature[consumer_op_name] = consumer->nd_sbp_signature();
+      if (op_name2op_conf.find(consumer_op_name) == op_name2op_conf.end()) {
+        op_name2op_conf[consumer_op_name] = consumer->op().op_conf();
+      }
+      OperatorConf& consumer_op_conf = op_name2op_conf.at(consumer_op_name);
+      for (const std::string& ibn : consumer->op().input_bns()) {
+        if (consumer->op().BnInOp2Lbi(ibn) == parallel_cast_out_lbi) {
+          const auto& new_val = GenLogicalBlobName(parallel_cast_in_lbi);
+          const auto& old_val = ReplaceInputLbnInOpCustomizedConf(&consumer_op_conf, ibn, new_val);
+          CHECK_EQ(GenLogicalBlobName(parallel_cast_out_lbi), old_val);
+        }
+      }
+    }
+    del_op_names.emplace_back(op_conf.name());
+    del_op_name_set.insert(op_conf.name());
+    VLOG(3) << "\tremove " << op_conf.name();
```

Comment:
比如placement改变就不能删除

Reply:
可以打印一下，信息多一些总是有好处的

Reply:
已添加，和删除打印删除的 node 是同一个 log 等级：

```
I20220927 14:41:36.282766 4021381 auto_parallel.cpp:74] Remove parallel cast ops for auto_parallel:
I20220927 14:41:36.397240 4021381 auto_parallel.cpp:154] 	remove model.bert.embeddings-hierarchical_parallel_cast-11
I20220927 14:41:36.397662 4021381 auto_parallel.cpp:117] hierarchical_parallel_cast-691 (with placement: oneflow.placement(type="cpu", ranks=[0])) can't be pruned, because producer model.cls_head.loss_func-broadcast_div-300's placement is oneflow.placement(type="cuda", ranks=[0, 1, 2, 3])
I20220927 14:41:36.397722 4021381 auto_parallel.cpp:117] hierarchical_parallel_cast-692 (with placement: oneflow.placement(type="cpu", ranks=[0])) can't be pruned, because producer model.cls_head.loss_func-scalar_mul-307's placement is oneflow.placement(type="cuda", ranks=[0, 1, 2, 3])
```